src/probnum/diffeq/odesolver.py

```diff
@@ -3,18 +3,103 @@
 """
 
 from abc import ABC, abstractmethod
+import warnings
+
+from probnum.diffeq.odesolution import ODESolution
 
 
 class ODESolver(ABC):
     """
     Interface for ODESolver.
     """
 
-    @abstractmethod
-    def solve(self, ivp, minstep, maxstep, **kwargs):
+    def __init__(self, ivp):
+        self.ivp = ivp
+        self.num_steps = 0
+
+    def solve(self, firststep, steprule, **kwargs):
         """
-        Every ODE solver has a solve() method.
-        Optional: callback function. Allows  e.g. printing variables
-        at runtime.
+        Solve an IVP.
+
+        Parameters
+        ----------
+        firststep : float
+            First step for adaptive step size rule.
+        steprule : StepRule
+            Step-size selection rule, e.g. constant steps or adaptive steps.
+        """
+        t, current_rv = self.initialise()
+        times, rvs = [t], [current_rv]
+        stepsize = firststep
+
+        while t < self.ivp.tmax:
+
+            t_new = t + stepsize
+            proposed_rv, errorest = self.step(t, t_new, current_rv, **kwargs)
+
+            if steprule.is_accepted(stepsize, errorest):
+                self.num_steps += 1
+                self.pre_accepted_callback(
+                    time=t_new, current_guess=proposed_rv, current_error=errorest
+                )
+                t = t_new
+                current_rv = proposed_rv
+                times.append(t)
+                rvs.append(current_rv)
+
+            suggested_stepsize = self._suggest_step(stepsize, errorest, steprule)
+            stepsize = min(suggested_stepsize, self.ivp.tmax - t)
+
+        odesol = self.postprocess(times=times, rvs=rvs)
+        return odesol
+
+    def _suggest_step(self, step, errorest, steprule):
+        """
+        Suggests step according to steprule and warns if
+        step is extremely small.
+
+        Raises
+        ------
+        RuntimeWarning
+            If suggested step is smaller than :math:`10^{-15}`.
         """
+        step = steprule.suggest(step, errorest)
+        if step < 1e-15:
+            warnmsg = "Stepsize is num. zero (%.1e)" % step
+            warnings.warn(message=warnmsg, category=RuntimeWarning)
+        return step
+
+    @abstractmethod
+    def initialise(self):
+        """Returns t0 and y0 (for the solver, which might be different to ivp.y0)"""
         raise NotImplementedError
+
+    @abstractmethod
+    def step(self, start, stop, current, **kwargs):
+        """Every ODE solver needs a step() method that returns a new random variable and an error estimate"""
+        raise NotImplementedError
+
+    def postprocess(self, times, rvs):
```

Comment:
What would you think about splitting `postprocess(times, rvs) -> ODESolution` into `postprocess(times, rvs) -> (times, rvs)` and `make_solution(times, rvs) -> ODESolution`? This goes together with my comment on the `postprocsess` chagen in `ivpfiltsmooth.py`.

Reply:
I also find this more explicit but still did it this way because smoothing takes a KalmanPosterior already so algorithmically it is simpler to make an odesolution and then smooth which would not quite work with splitting them. I am not sure in general, though... Also, it is not clear to me whether postprocess should then be placed before or after making the ODEsolution...

Reply:
I guess that second point is the real question. Should `postprocess` process the solution, or the saved values? Both seems reasonable. I guess for both cases we could make the naming more explicit (only if it actually helps of course)

Reply:
lets keep it as it is and keep in mind that the question still exists? I can make an 'improvement/refactoring' issue.

Reply:
That's also fine for me